DataFrame: Handle nanosecond-precision timestamp fields

gabor opened the pull request:
requires https://github.com/grafana/grafana-plugin-sdk-go/pull/647

- we modify the typescript dataframe-field type, we add `nanos`
- we modify the json-to-dataframe to handle the `nanos` field

(NOTE: an example usage is in https://github.com/grafana/grafana/pull/64687)

leeoniya commented on packages/grafana-data/src/types/dataFrame.ts:
```diff
@@ -136,6 +136,7 @@ export interface Field<T = any, V = Vector<T>> {
    */
   config: FieldConfig;
   values: V; // The raw field values
+  additionalValues?: Record<string, V>;
```
personally not a huge fan of long prop names, maybe just `values2` even if just to keep them adjacent in code/debugger due to same prefix?

i still think that just having `values: [1678387711084], values2: [0.000001]` or `values: [1678387711084], fracts: [0.000001]` is something to consider. it keeps the units uniform for the whole field and removes an extra unit-defining `nanos` layer/key.

gabor replied:
hi @leeoniya , will work on a better fieldName, 👍 

regarding the fractions, i totally see your point, but.. i tested it, and the floating-point numbers do not roundtrip unfortunately 😿 . try this:
```js
>>> console.log(1000000*0.000123)
123.00000000000001
```
and i need to do something like this because i need to get to a "whole nanosecond-number in a string" representation.

leeoniya replied:
you can Math.round() that result, but it looks like we're moving forward to int `nanos`, so probably no longer relevant.

ryantxu commented on packages/grafana-data/src/types/dataFrame.ts:
```diff
@@ -136,6 +136,7 @@ export interface Field<T = any, V = Vector<T>> {
    */
   config: FieldConfig;
   values: V; // The raw field values
+  additionalValues?: Record<string, V>;
```
Given that this is a pattern we know applies to time, and struggle to think of good cases for other things should we do something liket:
```suggestion
  /** When type === FieldType.time, we can support nanosecond time bla bla bla... 8?
  nanos?: number[];
```

gabor replied:
can do 👍 .. just to doublecheck, we can go directly to a javascript-array, no need to have an `ArrayVector` ?

gabor commented:
@ryantxu @leeoniya , i updated the PR:
- field is now called `nanos` and has the nubmers as an array
- added support to `dataframeToJSON` to make the cycle complete :)

gabor committed:
handle nanosec-precision timestamps

andresmgot commented on packages/grafana-data/src/dataframe/DataFrameJSON.ts:
```diff
@@ -188,16 +195,24 @@ export function dataFrameFromJSON(dto: DataFrameJSON): DataFrame {
       config: f.config ?? {},
       values: new ArrayVector(buffer),
       // the presence of this prop is an optimization signal & lookup for consumers
       entities: entities ?? {},
```
nit: since you are not using `nanos` elsewhere it can be defined here inline: `nanos: data?.nanos?.[index],`

gabor replied:
yes, i agree. the question is this... the `nanos` field will always be there that way, but it's value will be `undefined`. in other words, with current code:
```js
{ a: 1, b: 2}
```

inline nanos code:
```js
{ a: 1, b: 2, nanos: undefined }
```

what i do not know, do we care about such details of not having a field "be there but value will be undefined"?

andresmgot commented on packages/grafana-data/src/dataframe/DataFrameJSON.ts:
```diff
@@ -216,18 +231,34 @@ export function dataFrameToJSON(frame: DataFrame): DataFrameJSON {
       delete (sfield as any).entities;
       data.values.push(values.toArray());
+
+      if (nanos != null) {
```
the type of `nanos` is `number[] | undefined`, either the type or this condition is wrong

leeoniya replied:
`!= null` covers both `null` and `undefined` with fewer chars :)

andresmgot replied:
TIL _javascript magic_

```
> undefined != null
false
> undefined === null
false
> undefined == null
true
```

Anyway, it's just a `nit` but if `nanos` can be null I would define the type as `nanos?: number[] | null`

gabor replied:
sorry @andresmgot , it's an annoying trick (the `!= null`) and i in general really dislike such tricks, but this one is so good at basically `isNullish(value)` 😿

andresmgot reviewed (approved):
:+1: Code LGTM, I only found a couple of minor things. You can ignore the "breaking change" notice.

leeoniya reviewed (approved):
LGTM :+1: 

some notes:

the title of the PR/merge should likely not be "Logs:", but "DataFrame:"

> - we modify the typescript dataframe-field type, we add `additionalValues`

this description looks outdated relative to the current code
